### ISSUE TITLE
Problem when modifying tx.allowed_methods & searched for a SecRuleUpdateOperatorById

### ISSUE THREAD

**Issue by quenenni** [SpiderLabs/ModSecurity#755]
Hello,

I have a wordpress plugin that uses the DELETE method.
By default, it's blocked by modsec vie this rule:
```
SecRule REQUEST_METHOD "!@within %{tx.allowed_methods}" "phase:2,t:none,block,msg:'Method is not allowed by policy', severity:'2',id:'960032',..."
```

I tried several things:
* In modsecurity_crs_10_setup.conf, I modified the 'tx.allowed_methods' var to add 'DELETE':
```
setvar:'tx.allowed_methods=GET HEAD POST OPTIONS DELETE', \
```
But I still didn't work.

* I then changed the rule from 'base_rules/modsecurity_crs_30_http_policy.conf' to:
```
SecRule REQUEST_METHOD "!@within %{tx.allowed_methods}|DELETE"  "..."
```
And it worked.

My questions are:
* Is it normal it didn't work with my first method?
* I searched for a directive to update the operator of a rule so I hadn't to change the original file.
I found several update directives (SecRuleUpdateActionById, SecRuleUpdateTargetById, ..), but none that can update the operator.
Is there one I missed?

Thanks

**Comment by quenenni** [comment]
Sorry, I was not really efficient with this.
I think we can close this ticket as I won't have time to test more.

As it seems I'm the only one to have had that problem, I'm not sure it's a problem.

Again, sorry for the delay.

Good luck with your project.

**Comment by finchss** [comment]
I am facing a very similar issue:

in base_rules/modsecurity_crs_10_setup.conf:
```

SecAction \
  "id:'900012', \
  phase:1, \
  t:none, \
  setvar:'tx.allowed_methods=GET HEAD POST OPTIONS', \
  setvar:'tx.allowed_request_content_type=application/x-www-form-urlencoded|multipart/form-data|text/xml|application/xml|application/x-amf|application/json', \
  setvar:'tx.allowed_http_versions=HTTP/0.9 HTTP/1.0 HTTP/1.1', \
  setvar:'tx.restricted_extensions=.asa/ .asax/ .ascx/ .axd/ .backup/ .bak/ .bat/ .cdx/ .cer/ .cfg/ .cmd/ .com/ .config/ .conf/ .cs/ .csproj/ .csr/ .dat/ .db/ .dbf/ .d$
  setvar:'tx.restricted_headers=/Proxy-Connection/ /Lock-Token/ /Content-Range/ /Translate/ /via/ /if/', \
  nolog, \
  pass"
```

If I curl localhost I get a 403 Forbidden and the logs show:



 Initialising transaction (txid AQAcAcAc5cAcAqAcAcRcAnAc).
 Transaction context created (dcfg 7f6b265a4978).
 Starting phase REQUEST_HEADERS.
 This phase consists of 57 rule(s).
 Recipe: Invoking rule 7f6b26595768; [file "/usr/local/nginx-modsec/conf/modsecurity.conf"] [line "23"] [id "200000"].
 Rule 7f6b26595768: SecRule "REQUEST_HEADERS:Content-Type" "@rx (?:text|application)/xml" "phase:1,auditlog,id:200000,t:none,t:lowercase,pass,nolog,ctl:requestBodyProcessor=XML"
 Rule returned 0.
 No match, not chained -> mode NEXT_RULE.
 Recipe: Invoking rule 7f6b26593480; [file "/usr/local/nginx-modsec/conf/modsecurity.conf"] [line "30"] [id "200001"].
 Rule 7f6b26593480: SecRule "REQUEST_HEADERS:Content-Type" "@rx application/json" "phase:1,auditlog,id:200001,t:none,t:lowercase,pass,nolog,ctl:requestBodyProcessor=JSON"
 Rule returned 0.
 No match, not chained -> mode NEXT_RULE.
 Recipe: Invoking rule 7f6b202896b0; [file "/usr/local/nginx-modsec/conf/base_rules/modsecurity_crs_40_generic_attacks.conf"] [line "119"] [id "950012"] [rev "2"].
 Rule 7f6b202896b0: SecRule "REQUEST_HEADERS:'/(Content-Length|Transfer-Encoding)/'" "@rx ," "log,phase:1,rev:2,ver:OWASP_CRS/2.2.9,maturity:9,accuracy:9,t:none,capture,block,msg:'HTTP Request Smuggling Attack.',id:950012,tag:OWASP_CRS/WEB_ATTACK/REQUEST_SMUGGLING,tag:WASCTC/WASC-26,tag:OWASP_TOP_10/A1,tag:PCI/6.5.2,severity:2,setvar:tx.msg=%{rule.msg},setvar:tx.anomaly_score=+%{tx.critical_anomaly_score},setvar:tx.%{rule.id}-OWASP_CRS/WEB_ATTACK/REQUEST_SMUGGLING-%{matched_var_name}=%{tx.0}"
 Rule returned 0.
 No match, not chained -> mode NEXT_RULE.
 Recipe: Invoking rule 7f6b20119510; [file "/usr/local/nginx-modsec/conf/base_rules/modsecurity_crs_30_http_policy.conf"] [line "33"] [id "960032"] [rev "2"].
 Rule 7f6b20119510: SecRule "REQUEST_METHOD" "!@within %{tx.allowed_methods}" "log,phase:1,t:none,block,msg:'Method is not allowed by policy',severity:2,rev:2,ver:OWASP_CRS/2.2.9,maturity:9,accuracy:9,id:960032,tag:OWASP_CRS/POLICY/METHOD_NOT_ALLOWED,tag:WASCTC/WASC-15,tag:OWASP_TOP_10/A6,tag:OWASP_AppSensor/RE1,tag:PCI/12.1,logdata:%{matched_var},setvar:tx.msg=%{rule.msg},setvar:tx.anomaly_score=+%{tx.warning_anomaly_score},setvar:tx.%{rule.id}-OWASP_CRS/POLICY/METHOD_NOT_ALLOWED-%{matched_var_name}=%{matched_var}"
 Transformation completed in 2 usec.
 Executing operator "!within" with param "%{tx.allowed_methods}" against REQUEST_METHOD.
 Target value: "GET"
 Operator completed in 5 usec.
 Setting variable: tx.msg=%{rule.msg}
 Resolved macro %{rule.msg} to: Method is not allowed by policy
 Set variable "tx.msg" to "Method is not allowed by policy".
 Setting variable: tx.anomaly_score=+%{tx.warning_anomaly_score}
 Recorded original collection variable: tx.anomaly_score = "0"
 Relative change: anomaly_score=0+
 Set variable "tx.anomaly_score" to "0".
 Setting variable: tx.%{rule.id}-OWASP_CRS/POLICY/METHOD_NOT_ALLOWED-%{matched_var_name}=%{matched_var}
 Resolved macro %{rule.id} to: 960032
 Resolved macro %{matched_var_name} to: REQUEST_METHOD
 Resolved macro %{matched_var} to: GET
 Set variable "tx.960032-OWASP_CRS/POLICY/METHOD_NOT_ALLOWED-REQUEST_METHOD" to "GET".
 Rule returned 1.
 Match, intercepted -> returning.
 Resolved macro %{matched_var} to: GET
 Access denied with code 403 (phase 1). Match of "within %{tx.allowed_methods}" against "REQUEST_METHOD" required. [file "/usr/local/nginx-modsec/conf/base_rules/modsecurity_crs_30_http_policy.conf"] [line "33"] [id "960032"] [rev "2"] [msg "Method is not allowed by policy"] [data "GET"] [severity "CRITICAL"] [ver "OWASP_CRS/2.2.9"] [maturity "9"] [accuracy "9"] [tag "OWASP_CRS/POLICY/METHOD_NOT_ALLOWED"] [tag "WASCTC/WASC-15"] [tag "OWASP_TOP_10/A6"] [tag "OWASP_AppSensor/RE1"] [tag "PCI/12.1"]
 Initialising logging
.....



So apparently %{tx.allowed_methods} doesn't get expanded/resolved to "GET HEAD ..." the logs doesn't say anything regarding this macro, but get the other resolved.
Now if I change 
SecRule REQUEST_METHOD "!@within %{tx.allowed_methods}" "phase:1,t:none,block,msg:'Method is not allowed by policy',severity:'2',rev:$
to
 SecRule REQUEST_METHOD "!@within GET" "phase:1,t:none,block,msg:'Method is not allowed by policy',severity:'2',rev:$

it works.

So what is wrong?
I am using the latest mod_security 2.9.1 with nginx 

I use: 
2016/08/31 17:09:27 [notice] 19776#0: ModSecurity for nginx (STABLE)/2.9.1 (http://www.modsecurity.org/) configured.
2016/08/31 17:09:27 [notice] 19776#0: ModSecurity: APR compiled version="1.5.1"; loaded version="1.5.1"
2016/08/31 17:09:27 [notice] 19776#0: ModSecurity: PCRE compiled version="8.35 "; loaded version="8.35 2014-04-04"
2016/08/31 17:09:27 [notice] 19776#0: ModSecurity: LIBXML compiled version="2.9.1"
2016/08/31 17:09:27 [notice] 19776#0: ModSecurity: StatusEngine call: "2.9.1,nginx,1.5.1/1.5.1,8.35/8.35 2014-04-04,(null),2.9.1,ab9367d2a5c35c80e2700fb59de8125d81337ade"


Thanks.

**Comment by csanders-git** [comment]
@finchss This seems like an OWASP CRS ticket https://github.com/SpiderLabs/owasp-modsecurity-crs. I am also fairly confident that it has been resolved in version 3.0.0rc1, can you confirm?
